File: internal/server/agent/reverify.go
package agent

import (
	"time"

	"github.com/gofiber/fiber/v2"
)

// Async comment reverify agent endpoints (spec: specs/COMMENT_ASYNC_REVERIFY.md, PR-A).
// The connector polls /reverify/claim for posts to re-check, re-opens each, searches the
// DOM for the comment by actor + normalized text, and reports the verdict to
// /reverify/result. A positive match makes the backend append the append-only correction.
// Agent-token authed; scoped to the connector's org + assigned account.

// agentReverifyClaim handles GET /api/agent/reverify/claim?limit=N.
func (h *Handler) agentReverifyClaim(c *fiber.Ctx) error {
	orgID, _ := c.Locals("agent_org_id").(int64)
	accountID, _ := c.Locals("agent_assigned_account_id").(int64)
	if orgID <= 0 {
		return c.Status(403).JSON(fiber.Map{"error": "agent is not scoped to an organization"})
	}
	limit := c.QueryInt("limit", 5)
	if limit <= 0 {
		limit = 5
	}
	if limit > 20 {
		limit = 20
	}
	jobs, err := h.db.Coordination().ClaimDueReverifies(c.Context(), orgID, accountID, time.Now(), limit)
	if err != nil {
		return c.Status(500).JSON(fiber.Map{"error": err.Error()})
	}
	return c.JSON(fiber.Map{"reverifies": jobs, "count": len(jobs)})
}

// agentReverifyResult handles POST /api/agent/reverify/result with body
// {"id": N, "found": bool, "comment_permalink": "...", "notes": "..."}.
func (h *Handler) agentReverifyResult(c *fiber.Ctx) error {
	orgID, _ := c.Locals("agent_org_id").(int64)
	if orgID <= 0 {
		return c.Status(403).JSON(fiber.Map{"error": "agent is not scoped to an organization"})
	}
	var body struct {
		ID               int64  `json:"id"`
		Found            bool   `json:"found"`
		CommentPermalink string `json:"comment_permalink"`
		Notes            string `json:"notes"`
	}
	if err := c.BodyParser(&body); err != nil || body.ID <= 0 {
		return c.Status(400).JSON(fiber.Map{"error": "id required"})
	}
	corrected, err := h.db.Coordination().ApplyReverifyResult(c.Context(), orgID, body.ID, body.Found, body.CommentPermalink, body.Notes)
	if err != nil {
		return c.Status(500).JSON(fiber.Map{"error": err.Error()})
	}
	return c.JSON(fiber.Map{"ok": true, "corrected": corrected})
}

File: local-connector-extension/src/reverify.js
// Async comment reverify — connector poller (spec: specs/COMMENT_ASYNC_REVERIFY.md, PR-A).
// Claims due reverify jobs from the backend, re-opens each target post in the existing
// logged-in FB tab, asks the content script to search for the comment, and reports the
// verdict back. The backend appends the append-only correction on a positive match. Runs
// AFTER the outbox poller in the heartbeat cycle so the two never drive the tab at once.
var THGReverify = globalThis.THGReverify || (() => {
  let processing = null;

  async function claim() {
    const res = await THGApi.agentFetch('/api/agent/reverify/claim?limit=3');
    return (res && res.reverifies) || [];
  }

  async function report(id, found, permalink, notes) {
    await THGApi.agentFetch('/api/agent/reverify/result', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        id,
        found: !!found,
        comment_permalink: permalink || '',
        notes: notes || '',
      }),
    });
  }

  async function reverifyOne(job) {
    const url = String((job && job.target_url) || '').trim();
    // Only a DEFINITIVE search verdict (result.ok===true) is reported. A missing target,
    // no tab, busy composer, or content error leaves the row PENDING for the next cycle —
    // we must never mark a comment not_found just because we couldn't look this time.
    if (!url) return;
    const state = await THGFacebookState.ensureFacebookTabVisible(url, { focus: false });
    const tabId = state && state.tabId;
    if (!tabId) return;

    let result;
    try {
      result = await chrome.tabs.sendMessage(tabId, { type: 'thg_reverify_comment', message: job });
    } catch (e) {
      return; // content not reachable — leave pending
    }
    if (!result || result.ok !== true) return; // busy / not-ready — leave pending
    await report(job.id, !!result.found, result.comment_permalink || '', result.notes || '');
  }

  async function processOnce() {
    const jobs = await claim();
    for (const job of jobs) {
      // Best-effort per job: a failure leaves the row pending for the next cycle.
      try { await reverifyOne(job); } catch (e) { /* swallow; next poll retries */ }
    }
  }

  // process is guarded like the outbox poller so overlapping heartbeats don't double-run.
  function process() {
    if (processing) return processing;
    processing = processOnce().finally(() => { processing = null; });
    return processing;
  }

  return { process };
})();
globalThis.THGReverify = THGReverify;
if (typeof module !== 'undefined' && module.exports) module.exports = THGReverify;
